amdfftw: undo a broken change and adjust flags for performance tuning

amd-toolchain-support opened the pull request:
1. Undo a change from https://github.com/spack/spack/pull/46503 which breaks the clang build of this package.
2. Disable the dynamic dispatcher by default and add suitable compiler flags

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

amd-toolchain-support committed:
Changes to improve AMDFFTW performance

Co-authored-by: vijay kallesh <Vijay-teekinavar.Kallesh@amd.com>

alalazo commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -207,7 +207,12 @@ def configure(self, spec, prefix):
         # Spack user can not directly use AMD_ARCH for this purpose but should
         # use target variable to set appropriate -march option in AMD_ARCH.
-        options.append(f"AMD_ARCH={optimization_flags(self.compiler, spec.target)}")
+        options.append(f"AMD_ARCH={optimization_flags(self.compiler, spec.target).split('=')[-1]}")
```
Ugh, I though this option was meant to inject flags into compilation. Instead I see that it just accepts a fixed set of values:

https://github.com/amd/amd-fftw/blob/cfcb67c1f225c807d67dd99ad4f566d0a3eaf0f2/CMakeLists.txt#L13-L33

Should we add:
```
requires("target=zen:")
```
to the recipe? That would avoid the current state, where the recipe would just fail to build on Intel chips, or older AMD chips, right? Instead it would fail in concretization.

amd-toolchain-support replied:
With the CFLAGS, the code path in the amdfftw build system will bypass the logic around AMD_ARCH.

alalazo replied:
Reading the `CMakeLists.txt` above, wouldn't this result in:
```
SET(AMD_ARCH "znver1" CACHE STRING "select AMD zen version for Clang toolchain")
```
being executed, and thus injecting:
```
-march=znver1 -mprefer-vector-width=XXX
```
during the build? In this case wouldn't this break on Intel chips? Was:
```
requires("target=zen:")
```
having some issue?

amd-toolchain-support replied:
We are still using autotools to build I believe, since the FFTWBase package inherits from autotools and we don't attempt to change that: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/fftw/package.py#L12

In the autotools configure script injection of custom CFLAGS will short-circuit all of the platform detection logic.  In which case the Spack-provided arch flags will make it through and be used.

We could constrain to zen* only, but we are trying to allow build+run across all x86_64 platforms to allow as wide an audience as possible to try our packages.

alalazo replied:
Thanks for the explanation. LGTM then.

amd-toolchain-support committed:
Merge branch 'develop' into aocl/amdfftw_perf_fix

amd-toolchain-support committed:
Removing logic around AMD_ARCH

With CFLAGS, the code path in the amdfftw build system will bypass the logic around AMD_ARCH.

amd-toolchain-support committed:
Fixed flake8 found errors

amd-toolchain-support committed:
Merge branch 'develop' into aocl/amdfftw_perf_fix

amd-toolchain-support committed:
Merge branch 'develop' into aocl/amdfftw_perf_fix
